client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/unified-design-picker.tsx

```diff
@@ -278,20 +278,13 @@ const UnifiedDesignPickerStep: Step = ( { navigation, flow, stepName } ) => {
 		[ site ]
 	);
 
-	const getBadge = (
-		themeId: string,
-		forcePremium: boolean,
-		tooltipHeader: string,
-		tooltipMessage: string
-	) => (
+	const getBadge = ( themeId: string, isLockedStyleVariation: boolean ) => (
 		<ThemeTypeBadge
 			canGoToCheckout={ false }
-			forcePremium={ forcePremium }
+			isLockedStyleVariation={ isLockedStyleVariation }
```

Comment:
There is a high level of property propagation in the current architecture. The unified-design-picker resolves the isLockedStyleVariation and then passes it to getBadge -> ThemeTypeBadge -> tooltip, which causes a tight level of coupling and increases the maintenance costs since, if/when the property needs to be changed, we will need to modify more than one part of the codebase, which goes against the Single Responsibility principle.

I know it's a bit out of scope since it isn't a product change, but a better approach would be to have individual badge components for each type: wpcom upgrade, wpcom, and wporg. We can convert the getBadge function from a builder method to a factory method that returns the corresponding badge. This way, we can reduce the overall complexity of the architecture, improve readability since the badge wouldn't need to perform as much logic, and if the conditions for showing one badge or another change, we wouldn't need to modify more than one file across the stack.

I'm not expecting immediate action on this, this comment just aims to surface and raise awareness of the current issues we have with the architecture of our codebase and why we have high maintenance costs.

Reply:
> The unified-design-picker resolves the isLockedStyleVariation and then passes it to getBadge -> ThemeTypeBadge -> tooltip, which causes a tight level of coupling and increases the maintenance costs

Totally agree.

> a better approach would be to have individual badge components for each type: wpcom upgrade, wpcom, and wporg

Actually, I think a simpler approach, which aligns better with how React apps are typically developed, would be to store the current selected style variation in the state, and read that state from the tooltip component, so we don't need to pass the info a prop.

Reply:
> Actually, I think a simpler approach, which aligns better with how React apps are typically developed, would be to store the current selected style variation in the state, and read that state from the tooltip component, so we don't need to pass the info a prop.

These suggestions are not mutually exclusive and could potentially be combined. We could implement the factory method for creating badge components and also lift the state up to reduce prop drilling. This would reduce complexity and increase the maintainability of the codebase.